<a href="https://colab.research.google.com/github/Bryan-eng-lng/Machine-Learning-Projects-/blob/main/Titanic.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##Lets look at data and analyse data

In [32]:
import pandas as pd


df = pd.read_csv("/content/The Titanic dataset.csv")


In [33]:
print(df.shape)

(1302, 10)


In [34]:
print(df.head())

    1       2         3                            4       5    6       7  \
0  sn  pclass  survived                          NaN  gender  age  family   
1   1       3         0                  Mr. Anthony    male   42       0   
2   1       3         0                  Mr. Anthony    male   42       0   
3   2       3         0        Master. Eugene Joseph    male    ?       2   
4   3       2         0  Abbott, Mr. Rossmore Edward     NaN  NaN       2   

       8         9         10  
0   fare  embarked       date  
1   7.55       NaN  01-Jan-90  
2   7.55       NaN  01-Jan-90  
3  20.25         S  02-Jan-90  
4     **         S  03-Jan-90  


In [35]:
df.head()

,1,2,3,4,5,6,7,8,9,10
0,sn,pclass,survived,NaN,gender,age,family,fare,embarked,date
1,1,3,0,Mr. Anthony,male,42,0,7.55,NaN,01-Jan-90
2,1,3,0,Mr. Anthony,male,42,0,7.55,NaN,01-Jan-90
3,2,3,0,Master. Eugene Joseph,male,?,2,20.25,S,02-Jan-90
4,3,2,0,"Abbott, Mr. Rossmore Edward",NaN,NaN,2,**,S,03-Jan-90


In [36]:
print(df.columns)

Index(['1', '2', '3', '4', '5', '6', '7', '8', '9', '10'], dtype='object')


In [37]:
print(df.isnull().sum())

1       0
2       0
3       0
4       1
5       1
6     257
7       2
8       2
9       6
10      0
dtype: int64


In [38]:
df['3'].value_counts()

,count
3,
0,801
1,500
survived,1


##Survived = 0 and Not Survived = 1

In [39]:
df['1'].value_counts()

,count
1,
1,2
873,1
872,1
871,1
870,1
...,...
429,1
428,1
427,1


In [40]:
df = df.drop(columns=['1'] , axis=1)

print(df.head())

        2         3                            4       5    6       7      8  \
0  pclass  survived                          NaN  gender  age  family   fare   
1       3         0                  Mr. Anthony    male   42       0   7.55   
2       3         0                  Mr. Anthony    male   42       0   7.55   
3       3         0        Master. Eugene Joseph    male    ?       2  20.25   
4       2         0  Abbott, Mr. Rossmore Edward     NaN  NaN       2     **   

          9         10  
0  embarked       date  
1       NaN  01-Jan-90  
2       NaN  01-Jan-90  
3         S  02-Jan-90  
4         S  03-Jan-90  


In [41]:
df.head()

,2,3,4,5,6,7,8,9,10
0,pclass,survived,NaN,gender,age,family,fare,embarked,date
1,3,0,Mr. Anthony,male,42,0,7.55,NaN,01-Jan-90
2,3,0,Mr. Anthony,male,42,0,7.55,NaN,01-Jan-90
3,3,0,Master. Eugene Joseph,male,?,2,20.25,S,02-Jan-90
4,2,0,"Abbott, Mr. Rossmore Edward",NaN,NaN,2,**,S,03-Jan-90


##2 - pclass  , 3 - survived , 4 - NaN , 5 - gender , 6 - age , 7 - family , 8 - fare , 9 - embarked , 10 - date

In [42]:
df['2'].value_counts()

,count
2,
3,700
1,323
2,278
pclass,1


##Feature Engineering

In [43]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1302 entries, 0 to 1301
Data columns (total 9 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   2       1302 non-null   object
 1   3       1302 non-null   object
 2   4       1301 non-null   object
 3   5       1301 non-null   object
 4   6       1045 non-null   object
 5   7       1300 non-null   object
 6   8       1300 non-null   object
 7   9       1296 non-null   object
 8   10      1302 non-null   object
dtypes: object(9)
memory usage: 91.7+ KB


In [44]:
df = df.drop(index=0)
print(df.head())

   2  3                            4       5    6  7      8    9         10
1  3  0                  Mr. Anthony    male   42  0   7.55  NaN  01-Jan-90
2  3  0                  Mr. Anthony    male   42  0   7.55  NaN  01-Jan-90
3  3  0        Master. Eugene Joseph    male    ?  2  20.25    S  02-Jan-90
4  2  0  Abbott, Mr. Rossmore Edward     NaN  NaN  2     **    S  03-Jan-90
5  3  1  Abbott, Mr. Rossmore Edward  female   35  2  20.25    S  04-Jan-90


In [45]:
df.rename(columns={'2': 'pclass', '3': 'survived', '4': 'name', '5': 'gender', '6': 'age', '7': 'family', '8': 'fare', '9': 'embarked', '10': 'date'}, inplace=True)
print(df.head())

  pclass survived                         name  gender  age family   fare  \
1      3        0                  Mr. Anthony    male   42      0   7.55   
2      3        0                  Mr. Anthony    male   42      0   7.55   
3      3        0        Master. Eugene Joseph    male    ?      2  20.25   
4      2        0  Abbott, Mr. Rossmore Edward     NaN  NaN      2     **   
5      3        1  Abbott, Mr. Rossmore Edward  female   35      2  20.25   

  embarked       date  
1      NaN  01-Jan-90  
2      NaN  01-Jan-90  
3        S  02-Jan-90  
4        S  03-Jan-90  
5        S  04-Jan-90  


In [46]:
df = df.drop(columns=['name'], index=1)

print(df.head())


  pclass survived  gender  age family   fare embarked       date
2      3        0    male   42      0   7.55      NaN  01-Jan-90
3      3        0    male    ?      2  20.25        S  02-Jan-90
4      2        0     NaN  NaN      2     **        S  03-Jan-90
5      3        1  female   35      2  20.25        S  04-Jan-90
6      3        1  female   16      0   7.65        S  05-Jan-90


In [47]:
df.isnull().sum()

,0
pclass,0
survived,0
gender,1
age,257
family,2
fare,2
embarked,5
date,0


In [48]:
df['gender'] = df['gender'].fillna(df['gender'].mode()[0])

In [49]:
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()

df['gender'] = le.fit_transform(df['gender'])

In [50]:
df['age'] = pd.to_numeric(df['age'], errors='coerce')
df['age'] = df['age'].fillna(df['age'].mean())

In [51]:
df['fare'].value_counts()

,count
fare,
13,59
8.05,58
7.75,55
26,50
7.8958,49
...,...
9.325,1
7.7792,1
**,1


In [52]:
df['fare'] = pd.to_numeric(df['fare'], errors='coerce')
df['fare'] = df['fare'].fillna(df['fare'].mean())

In [53]:
df['embarked'].value_counts()

,count
embarked,
S,905
C,267
Q,123


In [54]:
df['embarked'] = df['embarked'].fillna(df['embarked'].mode()[0])

In [55]:
df['embarked'] = le.fit_transform(df['embarked'])

print(df.head())

  pclass survived  gender        age family      fare  embarked       date
2      3        0       1  42.000000      0   7.55000         2  01-Jan-90
3      3        0       1  29.967294      2  20.25000         2  02-Jan-90
4      2        0       1  29.967294      2  33.48216         2  03-Jan-90
5      3        1       0  35.000000      2  20.25000         2  04-Jan-90
6      3        1       0  16.000000      0   7.65000         2  05-Jan-90


In [56]:
df = df.drop(columns=['date'] , axis=1)

In [57]:
corr = df.corr()

corr

,pclass,survived,gender,age,family,fare,embarked
pclass,1.000000,-0.308653,0.121571,-0.364157,0.051201,-0.557384,0.187452
survived,-0.308653,1.000000,-0.528300,-0.051457,0.025424,0.242477,-0.177608
gender,0.121571,-0.528300,1.000000,0.062005,-0.188046,-0.184374,0.096733
age,-0.364157,-0.051457,0.062005,1.000000,-0.194858,0.169801,-0.073235
family,0.051201,0.025424,-0.188046,-0.194858,1.000000,0.226604,0.068687
fare,-0.557384,0.242477,-0.184374,0.169801,0.226604,1.000000,-0.239389
embarked,0.187452,-0.177608,0.096733,-0.073235,0.068687,-0.239389,1.000000


In [58]:
df = df.drop(columns=['family'] , axis=1)

In [59]:
df.head()

,pclass,survived,gender,age,fare,embarked
2,3,0,1,42.000000,7.55000,2
3,3,0,1,29.967294,20.25000,2
4,2,0,1,29.967294,33.48216,2
5,3,1,0,35.000000,20.25000,2
6,3,1,0,16.000000,7.65000,2


##Using Resampling to balance data

In [60]:
from sklearn.utils import resample

df_majority = df[df['survived'] == '0']

df_minority = df[df['survived'] == '1']

df_minority_unsampled = resample(df_minority , replace=True , n_samples=len(df_majority) , random_state=42)

df_balanced = pd.concat([df_majority , df_minority_unsampled])

print(df_balanced)

    pclass survived  gender        age      fare  embarked
2        3        0       1  42.000000   7.55000         2
3        3        0       1  29.967294  20.25000         2
4        2        0       1  29.967294  33.48216         2
8        2        0       1  30.000000  24.00000         0
12       3        0       1  30.000000   7.25000         2
..     ...      ...     ...        ...       ...       ...
858      2        1       0  14.000000  30.07080         0
924      2        1       1  29.000000  13.85830         0
348      3        1       1  19.000000   8.05000         2
964      3        1       1  29.967294  22.35830         0
167      2        1       0  20.000000  26.00000         2

[1600 rows x 6 columns]


##Importing Dependencies

In [61]:
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score , confusion_matrix , classification_report

##Feature = pclass = private class , gender , age , fare , embarked our target is survived . We have to know Is this person survived or not based on these features

In [62]:
scaler = StandardScaler()

df[['age' , 'fare']] = scaler.fit_transform(df[['age' , 'fare']])

In [63]:
df.head()

,pclass,survived,gender,age,fare,embarked
2,3,0,1,0.935936,-5.000808e-01,2
3,3,0,1,0.000000,-2.551715e-01,2
4,2,0,1,0.000000,1.370224e-16,2
5,3,1,0,0.391457,-2.551715e-01,2
6,3,1,0,-1.086414,-4.981523e-01,2


In [64]:
X = df.drop(columns=['survived'] , axis=1)

y = df['survived']



In [65]:
X_train , X_test , y_train , y_test = train_test_split( X , y , test_size = 0.2 , random_state=42)

## Implementing LogisticRegression Model

In [66]:
model = LogisticRegression(
    class_weight='balanced'
)

In [67]:
model.fit(X_train , y_train)

LogisticRegression(class_weight='balanced')

In [68]:
y_pred = model.predict(X_test)
print(y_pred)

['0' '0' '1' '0' '0' '1' '0' '0' '1' '0' '1' '1' '1' '0' '1' '1' '0' '0'
 '0' '1' '0' '0' '0' '0' '0' '0' '1' '1' '1' '0' '0' '1' '0' '1' '1' '1'
 '0' '0' '1' '1' '1' '0' '1' '1' '0' '1' '1' '0' '0' '0' '1' '0' '1' '0'
 '1' '1' '1' '1' '0' '0' '0' '0' '1' '0' '0' '1' '0' '0' '0' '1' '1' '0'
 '1' '0' '0' '0' '0' '0' '0' '0' '1' '0' '0' '1' '1' '0' '1' '1' '1' '1'
 '1' '0' '0' '1' '0' '0' '0' '1' '1' '0' '0' '1' '0' '0' '0' '1' '1' '1'
 '1' '1' '0' '0' '0' '0' '0' '0' '0' '0' '0' '1' '0' '0' '0' '0' '0' '1'
 '1' '1' '0' '0' '0' '1' '0' '1' '0' '0' '0' '1' '0' '0' '1' '0' '0' '0'
 '1' '0' '0' '0' '0' '1' '0' '0' '0' '0' '1' '0' '0' '1' '0' '0' '0' '0'
 '0' '1' '0' '1' '0' '0' '1' '0' '0' '1' '0' '1' '0' '1' '0' '1' '1' '1'
 '0' '0' '1' '1' '0' '0' '1' '0' '1' '0' '0' '1' '0' '0' '1' '1' '1' '0'
 '0' '0' '0' '1' '0' '0' '0' '0' '1' '0' '0' '1' '1' '1' '0' '0' '1' '0'
 '0' '0' '1' '1' '1' '0' '1' '0' '0' '0' '0' '0' '1' '0' '1' '1' '1' '1'
 '0' '1' '0' '0' '1' '0' '0' '0' '1' '1' '0' '0' '0

In [69]:
accuracy = accuracy_score(y_test , y_pred)

print(accuracy)

0.7884615384615384


In [70]:
accuracy = accuracy_score(y_train , model.predict(X_train))

print(accuracy)

0.7634615384615384


In [71]:
matrix = confusion_matrix(y_test , y_pred)

print(matrix)

[[134  32]
 [ 23  71]]


In [72]:
report = classification_report(y_test , y_pred)

print(report)


              precision    recall  f1-score   support

           0       0.85      0.81      0.83       166
           1       0.69      0.76      0.72        94

    accuracy                           0.79       260
   macro avg       0.77      0.78      0.78       260
weighted avg       0.79      0.79      0.79       260



##Implementing Decision Tree Model

In [73]:
from sklearn.tree import DecisionTreeClassifier

dt_model = DecisionTreeClassifier(
    class_weight='balanced',
    random_state=42,
    max_depth=4,
    min_samples_split=2,
    min_samples_leaf=1
)


In [74]:
dt_model.fit(X_train , y_train)



DecisionTreeClassifier(class_weight='balanced', max_depth=4, random_state=42)

In [75]:
y_pred = dt_model.predict(X_test)
print(y_pred)

['0' '0' '1' '0' '0' '1' '0' '0' '1' '0' '1' '1' '1' '0' '1' '1' '0' '0'
 '0' '1' '0' '0' '0' '0' '0' '0' '1' '1' '1' '0' '0' '1' '0' '1' '1' '1'
 '0' '0' '1' '1' '1' '0' '1' '1' '0' '0' '1' '0' '0' '0' '1' '0' '0' '0'
 '1' '1' '1' '1' '0' '0' '0' '0' '0' '0' '0' '0' '0' '0' '0' '1' '1' '0'
 '1' '0' '0' '0' '0' '0' '0' '0' '0' '0' '0' '1' '1' '0' '0' '1' '1' '1'
 '1' '0' '0' '1' '0' '0' '0' '0' '1' '1' '0' '1' '0' '0' '0' '1' '1' '1'
 '1' '0' '0' '0' '0' '0' '0' '0' '0' '0' '0' '1' '0' '0' '0' '0' '0' '1'
 '1' '1' '0' '0' '0' '1' '0' '0' '0' '0' '0' '1' '0' '0' '1' '0' '0' '0'
 '1' '0' '0' '0' '0' '1' '0' '0' '0' '0' '1' '0' '0' '1' '0' '0' '0' '0'
 '0' '1' '0' '1' '0' '0' '0' '0' '0' '1' '0' '1' '0' '1' '0' '0' '1' '1'
 '0' '0' '0' '1' '0' '0' '1' '0' '1' '0' '0' '0' '0' '0' '1' '1' '1' '0'
 '0' '0' '0' '1' '0' '0' '0' '0' '1' '0' '0' '1' '1' '1' '0' '0' '1' '0'
 '0' '0' '1' '1' '1' '0' '1' '0' '0' '0' '0' '0' '1' '0' '0' '1' '1' '1'
 '0' '1' '1' '0' '1' '0' '0' '0' '1' '0' '0' '0' '0

In [76]:
accuracy2 = accuracy_score(y_test , y_pred)

print(accuracy2)

0.8269230769230769


In [77]:
matrix2 = confusion_matrix(y_test , y_pred)

print(matrix2)

[[146  20]
 [ 25  69]]


In [78]:
report2 = classification_report(y_test , y_pred)

print(report2)

              precision    recall  f1-score   support

           0       0.85      0.88      0.87       166
           1       0.78      0.73      0.75        94

    accuracy                           0.83       260
   macro avg       0.81      0.81      0.81       260
weighted avg       0.83      0.83      0.83       260



#Acheived Good Accuracy 83% so target is to find good balance between survived and not survived